### ISSUE TITLE
[web] Flutter web puts something to the browser history and breaks navigation

### ISSUE THREAD

**Issue by chingiz98** [flutter/flutter#70358]
## Steps to Reproduce

1. Run any flutter-web application.

**Expected results:** 

You don't have anything in browser history

**Actual results:**

There is page called "origin" in browser history even though I didn't push anything to the history stack. Because of such behaviour, I have problem to close flutter webView that have flutter web application running in it. (Because WebViewController's canGoBack() returns true, even though it should be false)

The similar behaviour was seen in this issue #59185

![Screen Recording 2020-11-12 at 08 48 56](https://user-images.githubusercontent.com/5148741/98906097-957fac00-24cd-11eb-9379-db6ce7690f52.gif)

flutter doctor -v
<details>

[✓] Flutter (Channel beta, 1.23.0-18.1.pre, on Mac OS X 10.15.7 19H2 x86_64, locale en-GB)
    • Flutter version 1.23.0-18.1.pre at /Users/chingiz/fvm/versions/beta
    • Framework revision 198df796aa (4 weeks ago), 2020-10-15 12:04:33 -0700
    • Engine revision 1d12d82d9c
    • Dart version 2.11.0 (build 2.11.0-213.1.beta)

[!] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/chingiz/Library/Android/sdk
    • Platform android-30, build-tools 30.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)
    ✗ Android licenses not accepted.  To resolve this, run: flutter doctor --android-licenses

[✓] Xcode - develop for iOS and macOS (Xcode 12.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.1, Build version 12A7403
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 1.8.0_242-release-1644-b3-6222593)

[✓] Connected device (2 available)
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome 86.0.4240.198

! Doctor found issues in 1 category.

</details>

**Comment by darshankawar** [comment]
@chingiz98 Tried a sample code on chrome. The default back button is disabled on it. 

<img width="484" alt="Screenshot 2020-11-12 at 2 43 01 PM" src="https://user-images.githubusercontent.com/67046386/98919867-62e9a980-24f5-11eb-840c-c5ba72ba1913.png">


In safari browser, although the back button is enabled, tapping on it takes me back to default browser screen as below:

<img width="1037" alt="Screenshot 2020-11-12 at 2 44 03 PM" src="https://user-images.githubusercontent.com/67046386/98919986-8f052a80-24f5-11eb-8399-2c4799bf1465.png">
<img width="1023" alt="Screenshot 2020-11-12 at 2 44 12 PM" src="https://user-images.githubusercontent.com/67046386/98919998-91678480-24f5-11eb-8d1e-76776561b94d.png">

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel beta, 1.23.0-18.1.pre, on Mac OS X 10.15.4 19E2269 x86_64,
    locale en-IN)
    • Flutter version 1.23.0-18.1.pre at /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 198df796aa (4 weeks ago), 2020-10-15 12:04:33 -0700
    • Engine revision 1d12d82d9c
    • Dart version 2.11.0 (build 2.11.0-213.1.beta)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.1

[✓] Connected device (3 available)
    • SM A260G (mobile) • 5200763ebcfa861f • android-arm    • Android 8.1.0 (API
      27)
    • Web Server (web)  • web-server       • web-javascript • Flutter Tools
    • Chrome (web)      • chrome           • web-javascript • Google Chrome
      86.0.4240.193

• No issues found!


```

</details>

Are there any specific steps or code to replicate the issue you are facing ?
Thanks.

**Comment by chingiz98** [comment]
@darshankawar have you tried run it on "Web Server (web)" and open localhost link as i'm doing it on gif image?

**Comment by darshankawar** [comment]
@chingiz98 
Is this the correct interpretation of the issue, which seems to be happening in Safari browser only ?

[70358.mov.zip](https://github.com/flutter/flutter/files/5530245/70358.mov.zip)

**Comment by chingiz98** [comment]
@darshankawar Actually, not sure what's happening in your case. But take a look at screen recording in attachment.
So the steps I did are: 
1. Compiled the application in android studio ("Web Server" option) 
2. Clicked on the link that appeared in the terminal
3. Application loaded (and you can see there is "origin" in the history)
4. If you navigate with browser back button in the flutter web app, you can see that it is working, until you press back on the home page (I think browser is trying to get back to the "origin"). The expected behavior in this case should be: 

1. Browser gets back to the favorites page 
OR 
2. Back button should be disable on home page.

But in my case back button is enabled. If I press it, nothing will happen in the application itself, but the navigation with with browser back and forward button is not going to work anymore.

Why this is a problem for me?

Because this flutter web application is running in the web-view inside another mobile application (on android and iOS), and when user opens it and makes back gesture (same as you press back button in the browser on desktop pc) web-view is expected to be closed, but instead of this user have to make back gesture twice (again, I think, that web-view thinks that there is "origin" page inside history). And the worst thing, as I mentioned before, if user make back gesture on the home page for the first time, the web-view (browser) navigation will stop working.

[screen_record_1.zip](https://github.com/flutter/flutter/files/5530653/screen_record_1.zip)

In google chrome there is the same thing (just launch it as I show in the next attachment):
[screen_record_2.zip](https://github.com/flutter/flutter/files/5530744/screen_record_2.zip)

And the same thing happening with the example flutter project (so it is not related to my project):
[screen_record_3.zip](https://github.com/flutter/flutter/files/5530800/screen_record_3.zip)

**Comment by darshankawar** [comment]
@chingiz98 
Followed your lead here and compiled from AS with `web server` option using latest master and clicked on the link from console. I don't see the `origin` as you mentioned and tapping on back button shows me the same page.

[70358.mov 2.zip](https://github.com/flutter/flutter/files/5531000/70358.mov.2.zip)

<details>
<summary>flutter doctor -v</summary>

```
[✓] Flutter (Channel master, 1.24.0-8.0.pre.208, on Mac OS X 10.15.4 19E2269
    darwin-x64, locale en-IN)
    • Flutter version 1.24.0-8.0.pre.208 at
      /Users/dhs/documents/Fluttersdk/flutter
    • Framework revision 15d2d8a875 (11 hours ago), 2020-11-11 20:04:03 -0500
    • Engine revision d97a81c889
    • Dart version 2.12.0 (build 2.12.0-31.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.0)
    • Android SDK at /Users/dhs/Library/Android/sdk
    • Platform android-30, build-tools 30.0.0
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.0.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 12.0.1, Build version 12A7300
    • CocoaPods version 1.9.3

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin version 46.0.2
    • Dart plugin version 193.7361
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6222593)

[✓] VS Code (version 1.50.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.15.1

[✓] Connected device (4 available)
    • iPhone SE (2nd generation) (mobile) • 6C85835D-FBFD-4AB3-8DE8-B4FAD35E5367
      • ios            • com.apple.CoreSimulator.SimRuntime.iOS-14-0 (simulator)
    • macOS (desktop)                     • macos
      • darwin-x64     • Mac OS X 10.15.4 19E2269 darwin-x64
    • Web Server (web)                    • web-server
      • web-javascript • Flutter Tools
    • Chrome (web)                        • chrome
      • web-javascript • Google Chrome 86.0.4240.193

• No issues found!


```

</details>

**Comment by chingiz98** [comment]
@darshankawar I think, you will see "origin" if you long press on back button. (this just indicates that we have something in local browser history). You already have the issue. Back button should not be enabled on the first page.

**Comment by TahaTesser** [comment]
Hi @chingiz98 
i can reproduce it on multiple channels 

Preview

![ezgif com-gif-maker](https://user-images.githubusercontent.com/48603081/99403076-1268b680-2910-11eb-8687-def2b835ccd6.gif)


<details> 
<summary>code sample</summary> 

```dart
import 'package:flutter/material.dart';

void main() => runApp(MyApp());

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Material App',
      theme: ThemeData.dark(),
      home: Home(),
    );
  }
}

class Home extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text('Material App Bar'),
      ),
      body: Center(
        child: Container(
          child: Text('Hello World'),
        ),
      ),
      floatingActionButton: FloatingActionButton(
        child: Icon(Icons.add),
        onPressed: () {},
      ),
    );
  }
}

```

</details>

<details> 
<summary>multiple flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel beta, 1.23.0-18.1.pre, on macOS 11.0.1 20B29 x86_64, locale
    en-GB)
    • Flutter version 1.23.0-18.1.pre at /Users/tahatesser/Code/flutter_beta
    • Framework revision 198df796aa (5 weeks ago), 2020-10-15 12:04:33 -0700
    • Engine revision 1d12d82d9c
    • Dart version 2.11.0 (build 2.11.0-213.1.beta)

 
[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.2)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.2, Build version 12B45b
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.51.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (3 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 10 (API 29)
    • Web Server (web) • web-server       • web-javascript • Flutter Tools
    • Chrome (web)     • chrome           • web-javascript • Google Chrome
      86.0.4240.198

• No issues found!
```

```bash
[✓] Flutter (Channel master, 1.24.0-8.0.pre.268, on macOS 11.0.1 20B29
    darwin-x64, locale en-GB)
    • Flutter version 1.24.0-8.0.pre.268 at
      /Users/tahatesser/Code/flutter_master
    • Framework revision e803b13f01 (8 hours ago), 2020-11-16 22:55:59 -0800
    • Engine revision f0e80100f7
    • Dart version 2.12.0 (build 2.12.0-49.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.2)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.2, Build version 12B45b
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.51.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (4 available)
    • RMX2001 (mobile) • EUYTFEUSQSRGDA6D • android-arm64  • Android 10 (API 29)
    • macOS (desktop)  • macos            • darwin-x64     • macOS 11.0.1 20B29
      darwin-x64
    • Web Server (web) • web-server       • web-javascript • Flutter Tools
    • Chrome (web)     • chrome           • web-javascript • Google Chrome
      86.0.4240.198

• No issues found!
```

</details>

**Comment by mdebbar** [comment]
This is caused by the legacy web integration with the old Flutter Navigator. Once you migrate your app to use [Navigator 2.0](https://medium.com/flutter/learning-flutters-new-navigation-and-routing-system-7c9068155ade) this issue will disappear.

**Comment by TahaTesser** [comment]
Hi @chingiz98 
Thank you @mdebbar! 
I can confirm, origin doesn't show in navigator 2.0 

Code Sample
https://gist.github.com/johnpryan/5ce79aee5b5f83cfababa97c9cf0a204

<details> 
<summary>flutter doctor -v</summary> 

```bash
[✓] Flutter (Channel master, 1.24.0-8.0.pre.321, on macOS 11.0.1 20B29
    darwin-x64, locale en-GB)
    • Flutter version 1.24.0-8.0.pre.321 at
      /Users/tahatesser/Code/flutter_master
    • Framework revision 05dadb0120 (11 hours ago), 2020-11-20 07:33:05 +0530
    • Engine revision a0da844845
    • Dart version 2.12.0 (build 2.12.0-50.0.dev)

[✓] Android toolchain - develop for Android devices (Android SDK version 30.0.2)
    • Android SDK at /Users/tahatesser/Code/sdk
    • Platform android-30, build-tools 30.0.2
    • ANDROID_HOME = /Users/tahatesser/Code/sdk
    • Java binary at: /Applications/Android
      Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 12.2)
    • Xcode at /Volumes/Extreme/Xcode.app/Contents/Developer
    • Xcode 12.2, Build version 12B45b
    • CocoaPods version 1.10.0

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 4.1)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build
      1.8.0_242-release-1644-b3-6915495)

[✓] VS Code (version 1.51.1)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.16.0

[✓] Connected device (3 available)
    • macOS (desktop)  • macos      • darwin-x64     • macOS 11.0.1 20B29
      darwin-x64
    • Web Server (web) • web-server • web-javascript • Flutter Tools
    • Chrome (web)     • chrome     • web-javascript • Google Chrome
      87.0.4280.67

• No issues found!

```

</details>

Closing,  If you disagree, please write in the comments and I will reopen it.
Thank you

**Comment by vladmazur** [comment]
Is it possible to fix it without migrating to Navigator 2.0? My app is using library for navigation (GetX), and I'm not sure when they will release an update with null safety and Nav 2.0.
This origin issue is annoying really.

**Comment by RafaRuiz** [comment]
I was about to write what @vladmazur wrote. Same issue here using GetX
